### ISSUE TITLE
Windows containers don't have Performance Counters in registry

### ISSUE THREAD

**Issue by RajeshBabuJangili** [oshi/oshi#1976]
Hi , I am using Windows Server 2019 container, and using below code to get "SystemCpuLoadTicks" from windows container but every time I am getting values  [0, 0, 0, 0, 0, 0, 0, 0],because of this every time CPU Usage reporting 100%, please suggest me why I am getting all 0s every time ?

```java
{   
    SystemInfo si = new SystemInfo();
    HardwareAbstractionLayer hal = si.getHardware();

    CentralProcessor cpuInfo = hal.getProcessor();
    int coreCount = cpuInfo.getPhysicalProcessorCount();
    int processorCount = cpuInfo.getLogicalProcessorCount();
    CentralProcessor.ProcessorIdentifier cpuId = cpuInfo.getProcessorIdentifier();
    Map<String, WindowsRawData.Cpu> cpusMap = ImmutableMap.of("CPU", WindowsRawData.Cpu.builder()
            .cpuVendor(cpuId.getVendor())
            .cpuName(cpuId.getName())
            .numCores(coreCount)
            .numLogicalProcessors(processorCount)
            .cpuMaxSpeedMhz(cpuId.getVendorFreq() / 1000000L)
            .build());
    long[] loadTicks = cpuInfo.getSystemCpuLoadTicks();
    log.debug("OSHI library output (SystemCpuLoadTicks): " + Arrays.toString(loadTicks));    

    long idleTicks = loadTicks[CentralProcessor.TickType.IDLE.getIndex()];
    long stealTicks = loadTicks[CentralProcessor.TickType.STEAL.getIndex()];
    long systemTicks = loadTicks[CentralProcessor.TickType.SYSTEM.getIndex()]
            + loadTicks[CentralProcessor.TickType.IRQ.getIndex()]
            + loadTicks[CentralProcessor.TickType.SOFTIRQ.getIndex()];
    long userTicks = loadTicks[CentralProcessor.TickType.USER.getIndex()];
    long timeStamp = idleTicks + stealTicks + systemTicks + userTicks;


    Map<String, WindowsRawData.CpuPerf> cpusPerfMap = ImmutableMap.of("_Total", WindowsRawData.CpuPerf.builder()
            .timestamp(timeStamp)
            .idleTicks(idleTicks)
            .stealTicks(stealTicks)
            .systemTicks(systemTicks)
            .userTicks(userTicks)
            .build());
    
    rawData.setCpusPerf(cpusPerfMap);
    log.debug("OSHI library output (cpusPerf): " + rawData.getCpusPerf());
}
```

Output :
> appdynamics-infraviz-win-xlqnf==> [extension-scheduler-pool-0] 06 Mar 2022 22:52:05,933 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]

> appdynamics-infraviz-win-xlqnf==> [extension-scheduler-pool-0] 06 Mar 2022 22:52:05,933 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}

And addition I am getting below Error in log, please check :

> appdynamics-infraviz-win-xlqnf==> [extension-scheduler-pool-4] 03 Mar 2022 03:08:02,993 ERROR HkeyPerformanceDataUtil - Unable to locate English counter names in registry Perflib 009. Counters may need to be rebuilt: 
> com.sun.jna.platform.win32.Win32Exception: The system cannot find the file specified.
> 	at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:948) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
> 	at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:923) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
> 	at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:897) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
> 	at oshi.driver.windows.registry.HkeyPerformanceDataUtil.mapCounterIndicesFromRegistry(HkeyPerformanceDataUtil.java:301) ~[oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.driver.windows.registry.HkeyPerformanceDataUtil.<clinit>(HkeyPerformanceDataUtil.java:69) ~[oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.driver.windows.registry.ProcessPerformanceData.buildProcessMapFromRegistry(ProcessPerformanceData.java:69) ~[oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.software.os.windows.WindowsOperatingSystem.queryProcessMapFromRegistry(WindowsOperatingSystem.java:333) ~[oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.util.Memoizer$1.get(Memoizer.java:87) [oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.software.os.windows.WindowsOperatingSystem.processMapToList(WindowsOperatingSystem.java:303) [oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.software.os.windows.WindowsOperatingSystem.queryAllProcesses(WindowsOperatingSystem.java:265) [oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.software.common.AbstractOperatingSystem.getProcesses(AbstractOperatingSystem.java:107) [oshi-core-5.7.4.jar:5.7.4]
> 	at oshi.software.os.OperatingSystem.getProcesses(OperatingSystem.java:271) [oshi-core-5.7.4.jar:5.7.4]
> 	at com.appdynamics.sim.agent.extensions.servers.model.windows.WindowsRawCollector.fillOshiData(WindowsRawCollector.java:139) [servers-22.1.0-3252.jar:?]
> 	at com.appdynamics.sim.agent.extensions.servers.model.windows.WindowsRawCollector.collectRawData(WindowsRawCollector.java:101) [servers-22.1.0-3252.jar:?]
> 	at com.appdynamics.sim.agent.extensions.servers.model.windows.WindowsRawCollector.collectRawData(WindowsRawCollector.java:47) [servers-22.1.0-3252.jar:?]
> 	at com.appdynamics.sim.agent.extensions.servers.model.Server.collectAndReport(Server.java:65) [servers-22.1.0-3252.jar:?]
> 	at com.appdynamics.sim.agent.extensions.servers.ServersDataCollector.run(ServersDataCollector.java:94) [servers-22.1.0-3252.jar:?]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_292]
> 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:1.8.0_292]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_292]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_292]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_292]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_292]
> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292]

**Comment by dbwiddis** [comment]
This SO question looks to have similar symptoms (but no answer yet):
https://stackoverflow.com/questions/68117938/performance-counters-registry-key-missing-in-windows-servercore-container-azure

**Comment by dbwiddis** [comment]
So there are two errors here:
1. The 0 values from the performance counters indicate Windows is failing to read those counter values.  I'm surprised there are no associated log messages with those, such as in #1811.
2. The stack trace is from the `HKeyPerformanceDataUtil` that attempts to read process and thread counter information directly from the registry.  This is failing, but it is used when fetching process (or thread) counter data, not CPU ticks.   I was going to suggest calling `GlobalConfig.get(GlobalConfig.OSHI_OS_WINDOWS_HKEYPERFDATA, false);` at the beginning of your code, which will bypass the registry and use the performance counter API directly, but the 0 values above seem to indicate that the API eventually calls these registry values anyway.

Bottom line: the counters are corrupt/missing in this container.  It looks like rebuilding them fixes the problem, but I don't know why they would be missing in the first place.

**Comment by RajeshBabuJangili** [comment]
Hi Daniel, Thank you for your response, we recommend to the customer what you suggested and keep you updated.

**Comment by dbwiddis** [comment]
If their container is different than the one in the issue the solution is likely different. Also advised would be to make sure they have the most recent version of everything....

**Comment by RajeshBabuJangili** [comment]
Customer team tried adding below, but still CPU usage is 100%

RUN cd \windows\system32
RUN lodctr /R

to rebuild counters, which completes successfully, but CPU usage is still 100% for resulting image


Please suggest us next steps.

**Comment by dbwiddis** [comment]
From [https://docs.microsoft.com/en-us/azure/azure-monitor/app/performance-counters](this page):
>   Windows provides a wide variety of [performance counters](https://docs.microsoft.com/en-us/windows/desktop/perfctrs/about-performance-counters) such as processor, memory, and disk usage statistics. You can also define your own performance counters. Performance counters collection is supported as long as your application is running under IIS on an on-premises host, or virtual machine to which you have administrative access. Though applications running as Azure Web Apps don't have direct access to performance counters, a subset of available counters are collected by Application Insights.

It appears a permissions issue may be the root cause.  In order to even attempt to work around that I will need a lot more information about which "subset" is available.

**Comment by jeffaholmes** [comment]
This is the Dockerfile being use with info on specific base image:
FROM mcr.microsoft.com/windows/servercore:ltsc2019

https://github.com/CiscoDevNet/appdynamics-docker-images/blob/master/appd-machine-agent-analytics/windows/Dockerfile

**Comment by dbwiddis** [comment]
Note from [this article from appdynamics](https://www.appdynamics.com/blog/engineering/the-challenges-of-app-monitoring-with-net-core/):
> Engineers who are used to debugging apps in a traditional Windows environment find that valuable tools like Event Tracing for Windows (ETW) and **performance counters are not consistently available.**

I simply don't know where to get the equivalent information inside a container.

This may be something you want to ask about on an appdynamics support forum to see what alternatives there are.

**Comment by dbwiddis** [comment]
It looks like [this site](https://docs.appdynamics.com/display/PRO14S/Enable+Monitoring+for+Windows+Performance+Counters) gives instructions on exposing Windows performance counters to inside AppDynamics.

You can look at the classes in the `oshi.driver.windows.perfmon` package to see what counters OSHI reads, and configure them.  https://github.com/oshi/oshi/tree/master/oshi-core/src/main/java/oshi/driver/windows/perfmon

**Comment by dbwiddis** [comment]
Closing this issue as the last comment gave instructions to resolve and there haven't been any further updates in 6 months.  Feel free to reopen if you are still experiencing problems.
